### PR TITLE
Make publishing tf optional, enable publishing PoseStamped

### PR DESCRIPTION
**Pull request by johuber**
If the output of cartographer should be used as an input to an additional sensor fusion method, using the published TF transform is inconvenient or in our specific use case even harmful because we don't want to add the raw cartographer output to our TF tree.

With this change it becomes optional to broadcast to /tf. Morever there is an option to publish the tracked frame pose as a PoseStamped message.
We add two new optional parameters:
- `publish_to_tf` if false no tf transform is broadcasted
-  `publish_tracked_pose` If set `true` a PoseStamped representing the position of the tracked pose w.r.t. map frame is published.

If default launchers and settings are used this PR causes no change in the behavior.

### CODE REVIEW

**Comment by ojura** [comment]
@johuber I usually do `tf:=tf_null` when I don't want a node publishing stuff to TF. The only problem is when you want to subscribe to tf, but not publish - in that case introducing an another option is justified.

File: cartographer_ros/cartographer_ros/node_options.cc
Comment (MichaelGrupp) [inline cartographer_ros/cartographer_ros/node_options.cc]: From a conceptual point of view, a pose covariance doesn't exist in Cartographer - a `PoseStamped` would be totally sufficient to describe the trajectory without TF. Adding constant parameters for the covariance to Cartographer's configuration therefore doesn't make sense, as these are specific to your type of filter pipeline and don't have any use in Cartographer.

But then, you can also just use a different TF topic name as @ojura suggested.

Comment (MichaelGrupp) [reply cartographer_ros/cartographer_ros/node_options.cc]: > On the other hand I thought covariance is very related to the some cartographer settings, e.g. as voxel size.

Of course parameters like voxel size contribute to a systematic uncertainty. But defining a fixed pose covariance seems like a special requirement. And since it isn't used in Cartographer, it can be safely done outside of it. 

> So do you suggest to just use a different tf topic and obtaining the pose by subscribing to that topic?

I don't think this works since we also subscribe to TF (forget what I initially wrote). Maybe you can play around with the [`published_frame`](https://google-cartographer-ros.readthedocs.io/en/latest/configuration.html) parameter to publish to a child frame that doesn't affect the rest of the tree? (not quite sure at the moment if this will work). If not, I'd use a `PoseStamped`.

Comment (johuber) [reply cartographer_ros/cartographer_ros/node_options.cc]: > Of course parameters like voxel size contribute to a systematic uncertainty. But defining a fixed pose covariance seems like a special requirement. And since it isn't used in Cartographer, it can be safely done outside of it.

Agreed, I'll remove the covariance part.

> I don't think this works since we also subscribe to TF (forget what I initially wrote). 

In my understanding tf is not actually used by cartographer, but static transforms (base_link->imu, base_link->lidar) are read directly from an URDF.
However, cartographer_ros makes use of the tf tree to transform from the `tracked_frame` to the `published_frame`. Hence, if we publish the `tracked_frame` it works fine. when using a different tf topic.

> Maybe you can play around with the `published_frame` parameter to publish to a child frame that doesn't affect the rest of the tree? (not quite sure at the moment if this will work). If not, I'd use a PoseStamped.

I tried that, but unfortunately it always leads to having either two separate trees (causes visualization issues and warnings) or having wrong information in the tree.

So my suggestion:
I will introduce a flag to publish a `PoseStamped` of the `tracked_frame`, but will remove the flag for disabling tf publishing, since this issue can be solved by remapping tf. (Although I still think it's unreasonable to crowd the TCP/IP stack with data we are not using..)

**Comment by gaschler** [comment]
There is already a "hidden feature" to turn off publishing to tf by setting pose_publish_period_sec = 0.
Does this solve it?
I'm always hesitant to add new settings because we already have so many.

Also, what Michael explained, Cartographer does not use covariance. There were some experiments about this in the past but even in long hallways the covariance output did't show interesting principal components.

**Comment by johuber** [comment]
@gaschler 

> There is already a "hidden feature" to turn off publishing to tf by setting pose_publish_period_sec = 0.
Does this solve it?
I'm always hesitant to add new settings because we already have so man

Does solve the problem witht the tf treet, but then I'd still need to introduce some form of pose publisher and introduce a new parameter that defines the rate of said publisher.
Also it would probably require renaming to something like `tf_publish_period_sec` and `pose_msg_publish_period_sec`, because the naming would be confusing otherwise.

**Comment by johuber** [comment]
I changed it to publish a PoseStamped instead of PoseWithCovarianceStamped now

**Comment by johuber** [comment]
is there still any interest in this change? Should I close the PR?

**Comment by ojura** [comment]
Did you manage to solve your use case with the technique discussed above (tf remapping)?

**Comment by johuber** [comment]
It is possible with the tf remapping solution, but I still believe it would be a nice feature if we also have traditional ROS-style message publishing. Because in our use case (and I believe in most others) we prefer to subscribe to a pose feedback in the form of a regular ROS-message.

If you're not interested in that you may close the PR.

**Comment by ojura** [comment]
Disclaimer: I am not a maintainer, so I can't make that decision :) It may not be the case that they are disinterested, I think the maintainers are busy with other stuff since there's unfortunately been a lull with the repo activity in the past few months.

**Comment by faizansid77** [comment]
Why is this still not merged?

**Comment by johuber** [comment]
> Why is this still not merged?

was never approved.. I just rebased it to master and now appearently will need another approval.

**Comment by rotu** [comment]
What’s the status of this PR? I could really use this functionality.

**Comment by MichaelGrupp** [comment]
@bochen87 This PR was done at Enway, as far as I can tell. Do you want to rebase, sign it etc so we can do a fresh review? Since the PR was stale for a long time unfortunately and this seems to be a feature that would be welcomed by the community.

**Comment by johuber** [comment]
> @bochen87 This PR was done at Enway, as far as I can tell. Do you want to rebase, sign it etc so we can do a fresh review? Since the PR was stale for a long time unfortunately and this seems to be a feature that would be welcomed by the community.
> 
> Also the PR message should be updated since we already decided against having a covariance.

Yes done at Enway by me. I think I will get to do this within this or next week, no new changes should be required

**Comment by johuber** [comment]
@MichaelGrupp ready for review, did you offer to do this?

basically made no further changes (just rebased). It was already approved a while ago, but then the approval was lost due to rebasing

File: cartographer_ros/cartographer_ros/node_options.cc
Comment (MichaelGrupp) [inline cartographer_ros/cartographer_ros/node_options.cc:49]: I would remove `_msg` from the option name, it doesn't provide additional value.

File: cartographer_ros/cartographer_ros/node.cc
Comment (MichaelGrupp) [inline cartographer_ros/cartographer_ros/node.cc:119]: Don't advertise if the feature is disabled by the node options.

File: cartographer_ros/cartographer_ros/node.h
Comment (MichaelGrupp) [inline cartographer_ros/cartographer_ros/node.h:191]: One small thing: Since the topic is called `tracked_pose`, I would simplify the publisher name to `tracked_pose_publisher`.

**Review by MichaelGrupp — CHANGES_REQUESTED** [review]
Besides my inline comments, please also add a description of this topic / parameter to the ROS API documentation, under the "published topics" section.
https://github.com/cartographer-project/cartographer_ros/blob/master/docs/source/ros_api.rst#published-topics

File: cartographer_ros/cartographer_ros/node_options.cc
Comment (johuber) [reply cartographer_ros/cartographer_ros/node_options.cc:49]: done

File: cartographer_ros/cartographer_ros/node.cc
Comment (johuber) [reply cartographer_ros/cartographer_ros/node.cc:119]: done

File: cartographer_ros/cartographer_ros/node.h
Comment (johuber) [reply cartographer_ros/cartographer_ros/node.h:191]: done

**Comment by MichaelGrupp** [comment]
Please fix the build.

File: docs/source/configuration.rst
Comment (MichaelGrupp) [inline docs/source/configuration.rst:110]: Double backticks are needed in RST, as you did in the other file.

Comment (MichaelGrupp) [reply docs/source/configuration.rst:110]: The other messages here in this file are links, so we should probably also add a link for PoseStamped.

Comment (johuber) [reply docs/source/configuration.rst]: I opted for
```
``geometry_msgs/PoseStamped`_
```
to be consistent with the rest of this file. But forgot the underscore before, sorry!

Comment (johuber) [reply docs/source/configuration.rst]: oh I only just realized what this link does. Will add that too!

Comment (MichaelGrupp) [reply docs/source/configuration.rst]: yes, RST is quite unituitive with simple inline links :smile: If in doubt, you can also build the html version locally using `sphinx-build docs/ out/` and check it.

Thank you

**Review by MichaelGrupp — APPROVED** [review]

**Pull request merged**
